src/Val.h

```diff
@@ -506,38 +388,92 @@ class ValManager {
 
 extern ValManager* val_mgr;
 
+class IntVal : public Val {
+public:
+	IntVal(bro_int_t v)
+		: Val(base_type(TYPE_INT)), int_val(v)
+		{}
+
+	bro_int_t UnderlyingVal() const	{ return int_val; }
+
+protected:
+	IntVal(bro_int_t v, TypePtr t)
+		: Val(std::move(t)), int_val(v)
+		{}
+
+	bro_int_t int_val;
+};
+
+class BoolVal final : public IntVal {
```

Comment:
Fixed with next update.

Reply:
Uh oh, I figured out why this is sort of necessary.  The problem is the use of things like `TYPE_INTERNAL_INT` and `TYPE_INTERNAL_DOUBLE`.  When `BoolVal` and such derive from things like `IntVal` or `DoubleVal`, then those work fine; when they don't, then the internal types don't work, since it's no longer viable to call `AsInt` or such on the objects.

I'd rather not in this batch of changes rip out all of the `TYPE_INTERNAL_X` stuff ...  Can you abide leaving in the derivations from `IntVal` etc. for now?

Reply:
Darn ...  I'd rather not do that temporarily either, that would mean another API break later, and one with the potential to leads to bugs  then (like if somebody did a cast between these classes based on the temporary hierarchy). No immediate idea, need to look at how this being used.

Reply:
Not sure you'll care for this, but here's a thought on how to move forward: we introduce subclasses like  `IntValImplementation`, which are documented as only to be used for internal implementation.  Then `IntVal`, `EnumVal`, and `BoolVal` each derive from that, and it's clear up-front that those visible types are distinct, and their commonality is just an implementation detail.

Reply:
Yeah, I could see that.

One alternative thought: I haven't looked too closely, but is it primarily the `InternalXYZ` methods that need to be accounted for here? If so, we could implement them with a `switch` on the `TypeTag`, doing the down-cast manually as needed for each one.

Reply:
Looking things over, I think that approach would be workable.  It does though seem to me uglier than `IntValImplementation`, as it would require replacing  what are currently half a dozen `if`-`else` sequences / small `switch` bodies with significantly larger `switch` bodies.  The LOC impact of `IntValImplementation` would be quite a bit lower, and better isolated.

Reply:
Not sure I agree with the "better isolated" part (it's changes across a header file vs inside the internal implementations of a few functions). Generally it'd be ok with me to go with the subclasses, but one more question before we go that route: I'm wondering if we should go ahead and deprecate everything relying on the internal types at this point (e.g., especially `AsInternalXYZ()`, `Type::InternalType` and also `InternalTypeTag` itself) so that we can easily remove them later. I think that would make sense, but it would then automatically require rewriting those methods now as well anyways.

Reply:
Per our offline discussion, I've gone ahead with the `IntValImplementation` etc. approach for now.  BTW, I looked into what would be required to remove the various `Internal` stuff.  It looks tractable, but not simple enough that I'd want to bite it off now.

I've pushed updates now that I believe address all of the pending items per the above discussions.